src/Control/Carrier/Resource.hs

```diff
@@ -39,37 +32,31 @@ unliftResource handler = runReader (UnliftIO handler) . runResourceC
 --   . runState @Int 1
 --   $ myComputation
 -- @
-runResource :: MonadUnliftIO m
-            => ResourceC m a
-            -> m a
-runResource r = withRunInIO (\f -> runUnlifting (UnliftIO f) r)
+runResource :: ResourceC m a -> m a
+runResource = runResourceC
 
-newtype ResourceC m a = ResourceC { runResourceC :: ReaderC (UnliftIO m) m a }
+newtype ResourceC m a = ResourceC { runResourceC :: m a }
   deriving (Alternative, Applicative, Functor, Monad, Fail.MonadFail, MonadFix, MonadIO, MonadPlus)
 
-instance MonadUnliftIO m => MonadUnliftIO (ResourceC m) where
-  askUnliftIO = ResourceC . ReaderC $ \(UnliftIO h) ->
-    withUnliftIO $ \u -> pure (UnliftIO $ \r -> unliftIO u (unliftResource h r))
-
 instance MonadTrans ResourceC where
-  lift = ResourceC . lift
+  lift = ResourceC
 
-runUnlifting :: UnliftIO m -> ResourceC m a -> IO a
-runUnlifting h@(UnliftIO handler) = handler . runReader h . runResourceC
+instance MonadUnliftIO m => MonadUnliftIO (ResourceC m) where
+  withRunInIO f = ResourceC (withRunInIO (\ runInIO -> f (runInIO . runResourceC)))
 
-instance (Carrier sig m, MonadIO m) => Carrier (Resource :+: sig) (ResourceC m) where
+instance (Carrier sig m, MonadUnliftIO m) => Carrier (Resource :+: sig) (ResourceC m) where
   eff (L (Resource acquire release use k)) = do
-    handler <- ResourceC ask
+    handler <- askUnliftIO
     a <- liftIO (Exc.bracket
-      (runUnlifting handler acquire)
-      (runUnlifting handler . release)
-      (runUnlifting handler . use))
+      (unliftIO handler acquire)
+      (unliftIO handler . release)
+      (unliftIO handler . use))
```

Comment:
That enables us to use `askUnliftIO` to get the handler directly.

IMO this raises the question of whether `Resource` is even useful: the `MonadUnliftIO` constraint alone offers everything that this instance supplies, so it really becomes a question of whether there are other useful carriers. `MonadBaseControl`, perhaps? `haskeline`’s `MonadException`?

Either way, we may want to evaluate the utility of having this in the library rather than a more general effectful interface.

Reply:
As discussed on Slack, one option is to use `MonadBaseControl` and `lifted-base`, implementing `MonadBaseControl` for every carrier. That’s tough to do and requires a whole lot of `-XTypeFamilies`. Another alternative would be to pull in `exceptions` and provide `MonadMask` instances for everything (implementing `generalBracket` to correctly handle state information given an exit result).